source/AI.cpp

```diff
@@ -3385,35 +3385,128 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
+					default:
+					case Preferences::BoardingPriority::proximity:
+						return [current](Ship &other) noexcept -> double
+						{
+							return current.DistanceSquared(other.Position());
+						};
+				}
+			}();
+
+			using shipValue = pair<Ship *, double>;
+			auto boardable = vector<shipValue>{};
+
+			auto fillBoardable = [&ship, &foundEnemy, &boardable, &strategy](Ship &other) noexcept
+			{
+				if(CanBoard(ship, other))
+					boardable.emplace_back(&other, strategy(other));
+			};
+
+			if(shift)
+			{
+				for(Ship *escort : governmentRosters[ship.GetGovernment()])
+					if(escort->GetSystem() == ship.GetSystem())
+						fillBoardable(*escort);
+			}
+			else
+			{
+				auto ships = GetShipsList(ship, true);
+				boardable.reserve(ships.capacity());
+				// First check if we can board enemy ships, then allies.
+				for(Ship *enemy : ships)
+					fillBoardable(*enemy);
+				// The current target is not considered by GetShipsList.
+				if(target)
+					fillBoardable(*const_cast<Ship *>(target.get()));
+
+				if(boardable.empty())
+				{
+					ships = GetShipsList(ship, false);
+					if(ships.capacity() > boardable.capacity())
+						boardable.reserve(ships.capacity());
+					for(Ship *ally : ships)
+						fillBoardable(*ally);
+				}
+			}
+
+			if(boardable.empty())
+				activeCommands.Clear(Command::BOARD);
+			else
+			{
+				sort(boardable.begin(), boardable.end(),
+					[&ship, boardingPriority](
+						const shipValue &lhs, const shipValue &rhs
+					)
+					{
+						// If their cost is the same, prefer the closest ship.
+						if(boardingPriority == Preferences::BoardingPriority::value && lhs.second == rhs.second)
+							return lhs.first->Position().DistanceSquared(ship.Position()) >
+								rhs.first->Position().DistanceSquared(ship.Position());
+						else
+							return boardingPriority == Preferences::BoardingPriority::proximity ?
+								lhs.second > rhs.second : lhs.second < rhs.second;
+					}
+				);
+
+				// If there is no valid target, pick a new one (it could be exploding).
+				if(!target || !CanBoard(ship, *target))
+					ship.SetTargetShip((boardable.back().first->shared_from_this()));
+				// The WAIT command means we go to the next ship in the list relative to the one currently selected.
+				else if(activeCommands.Has(Command::WAIT))
+				{
+					auto boardingTarget = find_if(boardable.begin(), boardable.end(),
+						[&target](const shipValue &lhs)
+						{
+							return lhs.first == target.get();
+						}
+					);
 
-					bool isEnemy = other->GetGovernment()->IsEnemy(ship.GetGovernment());
-					double d = other->Position().Distance(ship.Position());
-					if((isEnemy && !foundEnemy) || (d < closest && isEnemy == foundEnemy))
+					if(boardingTarget != boardable.end())
 					{
-						closest = d;
-						foundEnemy = isEnemy;
-						foundAnything = true;
-						ship.SetTargetShip(other);
+						if(boardingTarget == boardable.begin())
+							ship.SetTargetShip(boardable.back().first->shared_from_this());
```

Comment:
Shouldn't this be `++boardable`? (instead of `.back()`)

Also the side effect seems unnecessary, I think it looks better if you use `next()`.

Reply:
I had no idea such things existed

Reply:
Oh yeah bruh I shouldn't review stuff at 2am lmao